mypy/checkexpr.py

```diff
@@ -1495,7 +1495,7 @@ def infer_overload_return_type(self,
             # We try returning a precise type if we can. If not, we give up and just return 'Any'.
             if all_same_types(return_types):
                 return return_types[0], inferred_types[0]
-            elif all_same_types(erase_type(typ) for typ in return_types):
+            elif all_same_types([erase_type(typ) for typ in return_types]):
```

Comment:
Note that `all_same_types()` calls `list()` on its argument. Maybe skip that if it's already a `list`?